tools/pre_ci.py

```
@@ -0,0 +1,201 @@
+
+                    if result.returncode == 0:
+                        print(f"✅ {desc} completed successfully.")
+                        self.record_result(desc, True)
+                    else:
+                        print(
+                            f"❌ FATAL: '{desc}' failed with exit code "
+                            f"{result.returncode}"
+                        )
+                        self.record_result(desc, False)
+                        success_overall = False
+                except Exception as exc:
```

Ruff (0.15.11)

[warning] Do not catch blind exception: `Exception`

(BLE001)